src/dune_pkg/opam_solver.ml

```diff
@@ -0,0 +1,190 @@
+  |> OpamFilter.filter_deps ~build:true ~post:true ~test:pkg_is_local
+       ~doc:pkg_is_local ~dev:pkg_is_local ~default:false
+
+module Dir_context_for_dune = struct
+  module Dir_context = Opam_0install.Dir_context
+  module Base_context = Context_with_local_packages (Dir_context)
+
+  (* Wrapper of [Opam_0install.Dir_context] which exposes its environment *)
+  type t =
+    { base_context : Base_context.t
+    ; env : string -> OpamVariable.variable_contents option
+    }
+
+  let create packages_dir_path ~env ~local_packages ~prefer_oldest =
+    let base_context =
+      Base_context.create
+        ~base_context:
+          (Dir_context.create ~prefer_oldest
+             ~constraints:OpamPackage.Name.Map.empty ~env packages_dir_path)
+        ~local_packages
+    in
+    { base_context; env }
+
+  (* Override [filter_deps] so that local packages get test and doc dependencies *)
+  let filter_deps t =
+    let local_packages = Base_context.local_packages t.base_context in
+    filter_deps_env ~local_packages ~env:(fun pkg v ->
+        if
+          List.mem OpamPackageVar.predefined_depends_variables v
+            ~equal:OpamVariable.Full.equal
+        then None
+        else
+          match OpamVariable.Full.to_string v with
+          | "version" ->
+            Some
+              (OpamTypes.S
+                 (OpamPackage.Version.to_string (OpamPackage.version pkg)))
+          | x -> t.env x)
+
+  type rejection = Base_context.rejection
+
+  let pp_rejection = Base_context.pp_rejection
+
+  let user_restrictions { base_context; _ } =
+    Base_context.user_restrictions base_context
+
+  let candidates { base_context; _ } = Base_context.candidates base_context
+end
+
+module Switch_context_for_dune = struct
+  module Switch_context = Opam_0install.Switch_context
+  module Base_context = Context_with_local_packages (Switch_context)
+
+  (* Wrapper of [Opam_0install.Switch_context] which exposes its switch state *)
+  type t =
+    { base_context : Base_context.t
+    ; switch_state : OpamStateTypes.unlocked OpamStateTypes.switch_state
+    }
+
+  let create switch_state ~local_packages ~prefer_oldest =
+    let base_context =
+      Base_context.create
+        ~base_context:
+          (Switch_context.create ~prefer_oldest
+             ~constraints:OpamPackage.Name.Map.empty switch_state)
+        ~local_packages
+    in
+    { base_context; switch_state }
+
+  (* Override [filter_deps] so that local packages get test and doc dependencies *)
+  let filter_deps t =
+    let local_packages = Base_context.local_packages t.base_context in
+    filter_deps_env ~local_packages ~env:(fun pkg v ->
+        if
+          List.mem OpamPackageVar.predefined_depends_variables v
+            ~equal:OpamVariable.Full.equal
+        then None
+        else
+          let r = OpamPackageVar.resolve_switch ~package:pkg t.switch_state v in
+          if r = None then
+            OpamConsole.warning "Unknown variable %S"
```

Comment:
It should probably not warn to the Opam console.

Reply:
Oops yes you're right.